docker/Dockerfile

```diff
@@ -17,51 +30,118 @@ RUN apt-get update && apt-get install -y libgeos-dev \
 # Install the BCMath PHP extension.
 RUN docker-php-ext-install bcmath
 
-# Inherit from the official Drupal 10 image.
-FROM drupal:10.1
+# Setup dependencies and sources for composer installations.
+FROM baseimage as composer-file
 
-# Set the farmOS and composer project repository URLs and versions.
-ARG FARMOS_REPO=https://github.com/farmOS/farmOS.git
-ARG FARMOS_VERSION=3.x
-ARG PROJECT_REPO=https://github.com/farmOS/composer-project.git
-ARG PROJECT_VERSION=3.x
+# Set the COMPOSER environment variables.
+ENV COMPOSER_HOME=tmp \
+    # Set the COMPOSER_MEMORY_LIMIT environment variable to unlimited.
+    COMPOSER_MEMORY_LIMIT=-1 \
+    # Allow root to install plugins.
+    COMPOSER_ALLOW_SUPERUSER=1
+
+# Install apt dependencies.
+RUN apt-get update && apt-get install -y \
+    # Install git and unzip (needed by Composer).
+    git unzip
+
+# Add the build-farmOS.sh script.
+COPY build-farmOS.sh /usr/local/bin/
+
+# Setup composer file for non interactive installation.
+WORKDIR ${FARMOS_PATH}
+RUN build-farmOS.sh
+
+##
+# Create layer with farmOS sources.
+FROM composer-file as farmos-sources
+
+# Install sources.
+RUN composer install --no-dev
+
+# Set the version in farm.info.yml.
+RUN sed -i "s|version: 3.x|version: ${FARMOS_VERSION}|g" ${FARMOS_PATH}/web/profiles/farm/farm.info.yml
+
+##
+# Create layer with farmOS dev sources.
+FROM composer-file as farmos-dev-sources
+
+# Install sources.
+RUN composer install
+
+# Set the version in farm.info.yml
+RUN sed -i "s|version: 3.x|version: ${FARMOS_VERSION}|g" ${FARMOS_PATH}/web/profiles/farm/farm.info.yml
```

Comment:
It can be done using `composer update` instead of `composer install`. Since `composer install --no-dev` in `farmos-sources` stage already installed the latest versions of production dependencies, `composer update` in `farmos-dev-sources` won't do anything with them, but it will install missing dev dependencies and will update `composer.lock` to include them.
https://github.com/farmOS/farmOS/commit/dda987c43069c04e42cc5acd7a61ec367d845ebc
https://github.com/wotnak/farmOS/actions/runs/7307749361/job/19913727994